Implement `sequenceNextNode()` function useful for `flow analysis`.

achimbab opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement `sequenceNextNode()` function useful for `flow analysis`.


Detailed description / Documentation draft:

The `sequenceNextNode(timestamp_column, event_column, event1, event2, ...)` is an aggregate function that returns a value of next event that matched the events.
```
sequenceNextNode([descending_order])(timestamp, event_column, event1, event2, ... eventN)
return:
  event_column[next_index], if the pattern is matched and next value exists.
  null, if the pattern isn’t matched or next value doesn't exist.
```
 
For instance, it can be used when events are `A->B->C->E->F` and you want to know the event following `B->C`, which is `E`.

It is useful for `flow analysis`.

The query statement searching the event following `B->C` : 
```SQL
CREATE TABLE test_flow (dt DateTime, id int, action String) 
ENGINE = MergeTree() 
PARTITION BY toYYYYMMDD(dt) 
ORDER BY id;

INSERT INTO test_flow VALUES (1, 1, 'A') (2, 1, 'B') (3, 1, 'C') (4, 1, 'E') (5, 1, 'F');

SELECT id, sequenceNextNode(0)(dt, action, action = 'B', action = 'C') FROM test_flow GROUP BY id;
```

Thank you.

otheng03 committed:
Implement sequenceNextNode

otheng03 committed:
Refactoring

otheng03 committed:
Serialization

otheng03 committed:
Fix

otheng03 committed:
Fix

Commit:
Documentation for sequenceNextNode

otheng03 committed:
For some functional stateless tests (ubsan)

otheng03 committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into sequenceNextNode

achimbab commented:
I added a documentation.

otheng03 committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into sequenceNextNode

otheng03 committed:
Use 0x80000000 as unmatched event

otheng03 committed:
Correct testcases

achimbab commented:
It seems that the failed test is not related to the PR.

1. test_distributed_ddl_parallel/test.py::test_all_in_parallel
```
Code: 36. DB::Exception: Received from 172.18.0.6:9000. DB::Exception: There was an error on [n1:9000]: Code: 36, e.displayText() = DB::Exception: external dictionary 'slow_dict' not found (version 21.2.1.5869). Stack trace:
```

2. test_testkeeper_back_to_back/test.py::test_multitransactions
```
>       assert len(list(minio.list_objects(cluster.minio_bucket, 'data/'))) == FILES_OVERHEAD_PER_PART_WIDE + FILES_OVERHEAD
E       AssertionError: assert 54 == (15 + 1)
```

achimbab commented:
Probably, It is possible to reduce memory usage and network traffic by encoding data and timestamps.

otheng03 committed:
Fix paramters for nulls

otheng03 committed:
Add sequenceFirstNode

achimbab commented:
I added `sequenceFirstNode` to get the starting points.
When events arguments are omitted in `sequenceNextNode`, it is replaced with `sequenceFirstNode` automatically.

achimbab commented:
It seems that the failed tests are irrelevant.

@alexey-milovidov  could you help review this PR?

If any problem, please let me know.

vdimir commented:
@achimbab 
Sorry for long delay. Thank you for your PR!

First of all, I think that we need more details in description, about terminology and what do we call `event` and `chain`, how do we consider data in table (some special form of graph I suppose), maybe add some links to external resources about this concepts.

Also it's good to add general description to code in comments, more details about how generally algorithm works, what it stores in aggregate function data, does it store all Nodes in memory, why do we use arena allocator for them and so on.

achimbab commented:
@vdimir 
It's okay.

I implemented `sequenceNextNode` for [flow analysis](https://mixpanel.com/topics/user-flow-analysis/)

> event 

It refers to activities of users of websites.

> chain

The `chain` consists of the events  of users.

> data in table

In the below tables schema, the action column refers to the event. 
```SQL
CREATE TABLE test_flow (dt DateTime, id int, action String) 
ENGINE = MergeTree() 
PARTITION BY toYYYYMMDD(dt) 
ORDER BY id;
```

For example, 
| dt | id | action |
| -- | -- | ------- |
| 2021-01-01 00:00:00 | 1 | visit_our_shoppingmall |
| 2021-01-01 00:01:00 | 1 | sign_in |
| 2021-01-01 00:02:00 | 1 | browse_item |
| 2021-01-01 00:03:00 | 1 | add_to_cart |
| 2021-01-01 00:04:00 | 1 | buy_all_items |


> Also it's good to add general description to code in comments, more details about how generally algorithm works,

I will write comments as soon as possible.

> what it stores in aggregate function data

The aggregate function stores an array of `(value of event_column, timestamp, bitset of matched_eventN)`

> does it store all Nodes in memory, why do we use arena allocator for them and so on.

It uses arena allocator for efficiency as [GroupArray](https://github.com/ClickHouse/ClickHouse/blob/master/src/AggregateFunctions/AggregateFunctionGroupArray.h) does.

otheng03 committed:
Add some comments

otheng03 committed:
Fix registerAggregateFunctions.cpp

otheng03 committed:
Merge remote-tracking branch 'origin/master' into HEAD

achimbab commented:
@vdimir 

I added comments to the source code.
Feel free to comment if there are anything I have to do in order to merge this PR. Have a nice day.

achimbab commented on src/AggregateFunctions/registerAggregateFunctions.cpp:
```diff
@@ -100,9 +100,9 @@ void registerAggregateFunctions()
         registerAggregateFunctionAggThrow(factory);
         registerAggregateFunctionRankCorrelation(factory);
         registerAggregateFunctionMannWhitney(factory);
+        registerAggregateFunctionSequenceNextNode(factory);
```
In order to merge with master branch without conflict.

vdimir committed:
Add __restrict modifier to AggregateFunctionSequenceNextNode

vdimir committed:
Minor changes in AggregateFunctionSequenceNextNode

vdimir commented on src/AggregateFunctions/AggregateFunctionSequenceNextNode.h:
```diff
@@ -0,0 +1,462 @@
+    {
+        bool operator()(const Node * lhs, const Node * rhs) const
+        {
+            if constexpr (Descending)
```
Seems that there are little in-determinism, result may vary if there are several results with same time:
```
INSERT INTO test_sequenceNextNode_Nullable values ('1970-01-01 09:00:01', 1, 'A');
INSERT INTO test_sequenceNextNode_Nullable values ('1970-01-01 09:00:01', 1, 'B');
INSERT INTO test_sequenceNextNode_Nullable values ('1970-01-01 09:00:02', 1, 'C');
INSERT INTO test_sequenceNextNode_Nullable values ('1970-01-01 09:00:02', 1, 'X');

SELECT sequenceNextNode(0)(dt, action, action = 'A', action = 'B') AS next_node  FROM test_sequenceNextNode_Nullable;
```
It may be changed when merges happens, maybe we can make it deterministic? For example, always choose smallest `value` comparing as strings.

vdimir commented on src/AggregateFunctions/AggregateFunctionSequenceNextNode.h:
```diff
@@ -0,0 +1,462 @@
+            value[i] = Node::read(buf, arena);
+    }
+
+    inline UInt32 calculateJump(const Data & data, const UInt32 i, const UInt32 j) const
+    {
+        UInt32 k = 0;
+        for (; k < events_size - j; ++k)
+            if (data.value[i - j]->events_bitset & (1 << (events_size - 1 - j - k)))
+                return k;
+        return k;
+    }
+
+    /// This method returns an index of next node that matched the events.
+    /// It is one as referring Boyer-Moore-Algorithm.
+    UInt32 getNextNodeIndex(Data & data) const
+    {
+        if (data.value.size() <= events_size)
+            return 0;
+
+        data.sort();
+
+        UInt32 i = events_size - 1;
+        while (i < data.value.size())
+        {
+            UInt32 j = 0;
+            /// It checks whether the chain of events are matched or not.
+            for (; j < events_size; ++j)
+                if (!(data.value[i - j]->events_bitset & (1 << (events_size - 1 - j))))
+                    break;
+
+            /// If the chain of events are matched returns the index of result value.
+            if (j == events_size)
+                return i + 1;
+
+            i += calculateJump(data, i, j);
+        }
+
+        return 0;
+    }
```
This definitely require  verbose explanation about how algorithm works. If it is a part of other algorithm (Boyer-Moore) it need to be explained anyway, how it is used, and how is adapted.

vdimir replied:
@achimbab 
Have you added comment about this (I suppose core) part?

Also I have another question: did you consider reusing functionality of other `sequence*` functions? What main difference, that requires separate algorithm?
https://github.com/ClickHouse/ClickHouse/blob/d9e22ba646d9946a4c3e3abee39a1842b552dfc3/src/AggregateFunctions/AggregateFunctionSequenceMatch.h

achimbab replied:
@vdimir 
Sure, I added comment about this.
If it isn't enough, please let me know.
https://github.com/achimbab/ClickHouse/blob/pr_sequenceNextNode/src/AggregateFunctions/AggregateFunctionSequenceNextNode.h

achimbab replied:
@vdimir 
There are some differences between two functions.
1. The return type
2. sequenceNextNode has to store values of event_column.
3. sequenceNextNode can find the index of result value in  event_column.

And two algorithmes have a functional difference. For example:
| dt | id | action |
| -- | -- | -------- |
| 1970-01-01 09:00:01 |  1 | a      |
| 1970-01-01 09:00:02 |  1 | b      |
| 1970-01-01 09:00:03 |  1 | b      |
| 1970-01-01 09:00:04 |  1 | c      |

```SQL
SELECT
    id,
    sequenceMatch('(?1)(?2)')(dt, action = 'a', action = 'c') AS result
FROM test_seq
GROUP BY id

 id   result
  1        1 
```
sequenceMatch returns 1. The order of events are `a -> b -> b -> c`, but It does not take into account event `b` in the middle, so `a -> c` makes sense even though there is `b` in the middle.

```SQL
SELECT
    id,
    sequenceNextNode(0)(dt, action, action = 'a', action = 'c') AS result
FROM test_seq
GROUP BY id

 id   result
  1   ᴺᵁᴸᴸ   
```
sequenceNextNode returns NULL. It is more sensitive in case where there is interrupting `b`.

vdimir commented:
@achimbab 
I left comments, generally it looks good, but core algorithm not so clear, I left some comments, I think it need to be clarified in more details.

Also I made minor changes (mostly style) and pushed it.

achimbab commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
```diff
@@ -505,3 +505,56 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+
+## sequenceNextNode {#sequenceNextNode}
+
+Returns a value of next event that matched an event chain.
```
Because events can have the same timestamp, it can't be strictly increasing, but monotonically increasing.

In this case, the ordered chain is:
```
('1970-01-01 09:00:01',1,'A');
('1970-01-01 09:00:01',1,'A');
('1970-01-01 09:00:02',...);
...omitted...
```
Because timestamp is the first ordering criteria.

So that it produces `'A'` is expected.

The omitted events can be clarified by applying your recommendation that `always choose smallest value comparing as strings.`
```
('1970-01-01 09:00:01',1,'A');
('1970-01-01 09:00:01',1,'A');
('1970-01-01 09:00:02',1,'B');
('1970-01-01 09:00:02',1,'C');
('1970-01-01 09:00:02',1,'X');
```

That's great. I will fix it in order to choose smallest value comparing as strings.

achimbab commented on src/AggregateFunctions/AggregateFunctionSequenceNextNode.h:
```diff
@@ -0,0 +1,462 @@
+        using Comparator = typename SequenceNextNodeGeneralData<T, Node, Descending>::Comparator;
+
+        if (!data(place).sorted && !data(rhs).sorted)
+            std::stable_sort(std::begin(a), std::end(a), Comparator{});
```
The method `sort` of Data is a general sorting method. It doesn't considering whether sort is necessary or not so that I don't use the method `sort` in `merge`.
My considering is two cases.
- At the line 224~225, it perform sorting whole list when two lists are not sorted.
- The below lines 226~, it perform partial sorting for each lists, and perform `inplace_merge` to speed up.
I don't know better sorting algorithm. Please tell me some more details.

achimbab commented on src/AggregateFunctions/AggregateFunctionSequenceNextNode.h:
```diff
@@ -0,0 +1,462 @@
+        //   aggregated events: [A -> B -> C]
+        //   events to find: [C -> D]
+        //   [C -> D] is not matched to 'A -> B -> C' so that it returns null.
+        return std::make_shared<AggregateFunctionNullVariadic<false, false, true, true>>(nested_function, arguments, params);
```
I will fix it. Thank you.

achimbab commented on src/AggregateFunctions/AggregateFunctionSequenceNextNode.cpp:
```diff
@@ -0,0 +1,114 @@
+    for (const auto i : ext::range(2, argument_types.size()))
+    {
+        const auto * cond_arg = argument_types[i].get();
+        if (!isUInt8(cond_arg))
```
In some cases,  event chains refer to another columns.
For example, 
```
dt                                    id     action            page     
2021-02-12 00:00:01     1     click_btn1      main
2021-02-12 00:01:01     1     click_btn2      shopping
2021-02-12 00:02:01     1     click_btn3      news
2021-02-12 00:03:01     1     click_btn4      sports
```
```
SELECT sequenceNextNode(1)(dt, page, action = 'click_btn1', action = 'click_btn2', page = 'news') AS next_node  FROM test_table;
```
So that I created column with result of comparison.

otheng03 committed:
Add comments and fix AggregateFunctionSequenceNextNode

otheng03 committed:
Fix comparator of NodeString

otheng03 committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into HEAD

achimbab commented:
@vdimir 
I fixed source codes. and added comments.
It seems that the failed tests are irrelevant.

Commit:
Fix comparator of sequenceNextNode for more deterministic sorting

achimbab commented:
@vdimir 
I fixed some codes for more deterministic sorting

1. Remove EMPTY_EVENTS_BITSET
2. Remove event_bitsets from comparison criterion in order to more deterministic sorting.
3. Replace the type of bitset with std::bitset
4. Remove in-determinisms in testcases 

Thank you.

otheng03 committed:
Fix comparator

otheng03 committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into HEAD

vdimir commented on src/AggregateFunctions/AggregateFunctionSequenceNextNode.h:
```diff
@@ -0,0 +1,473 @@
+        bool operator()(const Node * lhs, const Node * rhs) const
+        {
+            if constexpr (Descending)
+                return lhs->event_time == rhs->event_time ? !lhs->compare(rhs) : lhs->event_time > rhs->event_time;
```
Query 
```
INSERT INTO test_sequenceNextNode_Nullable values ('1970-01-01 09:00:01',1,'A');
INSERT INTO test_sequenceNextNode_Nullable values ('1970-01-01 09:00:01',1,'A');

SELECT sequenceNextNode(1)(dt, action, action = 'A') AS next_node  FROM test_sequenceNextNode_Nullable;
```

leads to assertion error in debug mode `/contrib/libcxx/include/algorithm:814: _LIBCPP_ASSERT '!__comp_(__l, __r)' failed. Comparator does not induce a strict weak ordering`
We violate rule that `a < b` and `b < a` cannot return both `true`. https://stackoverflow.com/a/981299

I think we can leave condition `lhs->compare(rhs)` without negation, because `Decreasing` should affect only on time ordering, but if time is the same we can choose lowest string for decreasing and increasing.

How do you think?

vdimir committed:
Remove MAX_EVENTS_SIZE from AggregateFunctionSequenceNextNode.h

vdimir committed:
Fix strict weak ordering, return always lower string in sequenceNextNode

vdimir committed:
Minor changes, add comments

vdimir commented on src/AggregateFunctions/AggregateFunctionSequenceNextNode.h:
```diff
@@ -0,0 +1,483 @@
+    }
+
+    /// This method returns an index of next node that matched the events.
+    /// It is one as referring Boyer-Moore-Algorithm(https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_string-search_algorithm).
```
Important part of Boyer-Moore is shift table for bad character. But we do not build it explicitly because we know position of each event and can determinate it from flag in bitset. Is it right?

achimbab replied:
That's right.

vdimir reviewed (commented):
Ok, I think I got core idea of algorithm. It's not so sophisticated, but I found it's really clever to apply it to such kind of task in such way (with bitsets and so on).

I left some comments and made some minor changes, feel free to comment it too.

vdimir commented on src/AggregateFunctions/AggregateFunctionSequenceNextNode.h:
```diff
@@ -0,0 +1,462 @@
+    UInt64 size; // size of payload
+
+    DataTypeDateTime::FieldType event_time;
+    UInt32 events_bitset; // Bitsets of UInt32 are easy to compare. (< operator on bitsets)
```
We decided not to use compassion, compare strings instead, is it?

achimbab replied:
Yes, I removed the events_bitset from the comparator.

achimbab commented on src/AggregateFunctions/AggregateFunctionSequenceNextNode.h:
```diff
@@ -350,7 +362,7 @@ class SequenceFirstNodeImpl final
         const AggregateFunctionProperties &) const override
     {
-        return std::make_shared<AggregateFunctionNullVariadic<false, false, true, true>>(nested_function, arguments, params);
+        return std::make_shared<AggregateFunctionNullVariadic<true, false, true, true>>(nested_function, arguments, params);
```
if result_is_nullable is false,  it branches to the line 303 and the type of to is `ColumnNullable` so tests was OK.
I think `insertion_requires_nullable_column` doesn't need right?
I will remove the option `insertion_requires_nullable_column`.

```C++
279     void insertResultInto(AggregateDataPtr place, IColumn & to, Arena * arena) const override                               
280     {                                                                                                                       
281         if constexpr (result_is_nullable)                                                                                   
282         {                                                                                                                   
283             ColumnNullable & to_concrete = assert_cast<ColumnNullable &>(to);                                               
284             if (this->getFlag(place))                                                                                       
285             {                                                                                                               
286                 if constexpr (insertion_requires_nullable_column)                                                           
287                 {                                                                                                           
288                     this->nested_function->insertResultInto(this->nestedPlace(place), to_concrete, arena);                  
289                 }                                                                                                           
290                 else                                                                                                        
291                 {                                                                                                           
292                     this->nested_function->insertResultInto(this->nestedPlace(place), to_concrete.getNestedColumn(), arena);
293                     to_concrete.getNullMapData().push_back(0);                                                              
294                 }                                                                                                           
295             }                                                                                                               
296             else                                                                                                            
297             {                                                                                                               
298                 to_concrete.insertDefault();                                                                                
299             }                                                                                                               
300         }                                                                                                                   
301         else                                                                                                                
302         {                                                                                                                   
303             this->nested_function->insertResultInto(this->nestedPlace(place), to, arena);                                   
304         }                                                                                                                   
305     }                                                                                                                       
```

vdimir committed:
Update test sequence_next_node, choose lowest string for same timestamps

achimbab commented on src/AggregateFunctions/AggregateFunctionSequenceNextNode.h:
```diff
@@ -0,0 +1,483 @@
+namespace DB
+{
+
+/// NodeBase used to implement a linked list for storage of SequenceNextNodeImpl
```
the lined list will be stored in the value of SequenceNextNodeGeneralData.
The list stores values of event_column.

achimbab commented:
@vdimir 
I found an exceptional case.
So I will fix the core part of this aggregation function.
Sorry, I will make change as soon as possible for tomorrow.

vdimir commented:
> I found an exceptional case.

Do you mean that there are cases when algorithm provides wrong result? Could you give an example?

> So I will fix the core part of this aggregation function.
> Sorry, I will make change it as soon as possible.

Sure, it's up to you.

achimbab commented:
An exceptional case is:

data:

| timestamp | id | page |
| -- | -- | ------ |
| 1 | 1 | home |
| 2 | 1 | home |
| 3 | 1 | giftbox |
| 4 | 1 | wish |
| 1 | 2 | home |
| 2 | 2 | giftbox |
| 3 | 2 | wish |


1. Find first page.

```
select id, sequenceNextNode(0)(timestamp, page) from table group by id
```

```
home (2 people)
```
 
2. Get a next page of `home`:

```
select id, sequenceNextNode(0)(timestamp, page, page = 'home') from table group by id
```

```
home (2 people) -> home (1 person)
                -> giftbox (1 person)
```

3. Get a next page of `home` -> `giftbox`

```
select id, sequenceNextNode(0)(timestamp, page, page = 'home', page = 'giftbox') from table group by id
```

```
home (2 people) -> home (1 person)
                -> giftbox (1 person) -> wish (2 people)
```
Oops, there are 2 people in the `wish`, but there is only 1 person in the previous flow `giftbox`

I will fix it.

otheng03 committed:
Change the behavior of sequenceNextNode.

otheng03 committed:
Remove insertion_requires_nullable_column

otheng03 committed:
Fix special-build-failure and add argument validation

otheng03 committed:
Fix a bug in sequenceNextNode

achimbab commented:
I'm fixing this aggregation function.
I will notify you when I think it is finished.

otheng03 committed:
Add base condition

otheng03 committed:
Merge remote-tracking branch 'upstream/master' into pr_sequenceNextNode

otheng03 committed:
Fix reverse indexing of TAIL in serialize()

otheng03 committed:
Include bitset and rename test

otheng03 committed:
Merge remote-tracking branch 'origin/master' into HEAD

otheng03 committed:
Merge remote-tracking branch 'origin/master' into HEAD

achimbab commented:
@vdimir 
It seems that failed testcases are irrelavent with this PR.
May I get review again?

achimbab commented:
@vdimir 

I modified the core algorithm because, the core algorithm had the [exceptional case](https://github.com/ClickHouse/ClickHouse/pull/19766#issuecomment-795175943).

The changes from the last version.
- Deleted an algorithm like `Boyer-Moore`. The current version of the algorithm uses just simple for-loop in order to check that the chain of events are matched.
- Modified parameters
- Added `base condition` argument

### Modified parameters

- option 1: `direction` - Used to navigate to directions
  - forward : Moving forward
  - backward: Moving backward

- option 2: `base` - Used to set the base point.
  - head : Set the base point to the first event
  - tail : Set the base point to the last event
  - first_match : Set the base point to the first matched event1
  - last_match : Set the base point to the last matched event1

### Added `base condition` argument.
- `base_condition` — Condition that the base point must fulfill.

### Examples

**Behavior for `forward` and `head`**

```SQL
ALTER TABLE test_flow DELETE WHERE 1 = 1 settings mutations_sync = 1;
INSERT INTO test_flow VALUES (1, 1, 'Home') (2, 1, 'Gift') (3, 1, 'Exit');
INSERT INTO test_flow VALUES (1, 2, 'Home') (2, 2, 'Home') (3, 2, 'Gift') (4, 2, 'Basket');
INSERT INTO test_flow VALUES (1, 3, 'Gift') (2, 3, 'Home') (3, 3, 'Gift') (4, 3, 'Basket');
```

```SQL
SELECT id, sequenceNextNode('forward', 'head')(dt, page = 'Home', page, page = 'Home', page = 'Gift') FROM test_flow GROUP BY id;
 
                  dt   id   page
 1970-01-01 09:00:01    1   Home // Base point, Matched with Home
 1970-01-01 09:00:02    1   Gift // Matched with Gift
 1970-01-01 09:00:03    1   Exit // The result 
 1970-01-01 09:00:01    2   Home // Base point, Matched with Home
 1970-01-01 09:00:02    2   Home // Unmatched with Gift
 1970-01-01 09:00:03    2   Gift
 1970-01-01 09:00:04    2   Basket    
 
 1970-01-01 09:00:01    3   Gift // Base point, Unmatched with Home
 1970-01-01 09:00:02    3   Home      
 1970-01-01 09:00:03    3   Gift      
 1970-01-01 09:00:04    3   Basket    
```

**Behavior for `backward` and `tail`**

```SQL
SELECT id, sequenceNextNode('backward', 'tail')(dt, page = 'Basket', page, page = 'Basket', page = 'Gift') FROM test_flow GROUP BY id;
                 dt   id   page
1970-01-01 09:00:01    1   Home
1970-01-01 09:00:02    1   Gift
1970-01-01 09:00:03    1   Exit // Base point, Unmatched with Basket
                                     
1970-01-01 09:00:01    2   Home 
1970-01-01 09:00:02    2   Home // The result 
1970-01-01 09:00:03    2   Gift // Matched with Gift
1970-01-01 09:00:04    2   Basket // Base point, Matched with Basket
                                     
1970-01-01 09:00:01    3   Gift
1970-01-01 09:00:02    3   Home // The result 
1970-01-01 09:00:03    3   Gift // Base point, Matched with Gift
1970-01-01 09:00:04    3   Basket // Base point, Matched with Basket
```


**Behavior for `forward` and `first_match`**

```SQL
SELECT id, sequenceNextNode('forward', 'first_match')(dt, page = 'Gift', page, page = 'Gift') FROM test_flow GROUP BY id;
                 dt   id   page
1970-01-01 09:00:01    1   Home
1970-01-01 09:00:02    1   Gift // Base point
1970-01-01 09:00:03    1   Exit // The result
                                     
1970-01-01 09:00:01    2   Home 
1970-01-01 09:00:02    2   Home 
1970-01-01 09:00:03    2   Gift // Base point
1970-01-01 09:00:04    2   Basket  The result
                                     
1970-01-01 09:00:01    3   Gift // Base point
1970-01-01 09:00:02    3   Home // Thre result
1970-01-01 09:00:03    3   Gift   
1970-01-01 09:00:04    3   Basket    
```

```SQL
SELECT id, sequenceNextNode('forward', 'first_match')(dt, page = 'Gift', page, page = 'Gift', page = 'Home') FROM test_flow GROUP BY id;
                 dt   id   page
1970-01-01 09:00:01    1   Home
1970-01-01 09:00:02    1   Gift // Base point
1970-01-01 09:00:03    1   Exit // Unmatched with Home
                                     
1970-01-01 09:00:01    2   Home 
1970-01-01 09:00:02    2   Home 
1970-01-01 09:00:03    2   Gift // Base point
1970-01-01 09:00:04    2   Basket // Unmatched with Home
                                     
1970-01-01 09:00:01    3   Gift // Base point
1970-01-01 09:00:02    3   Home // Matched with Home
1970-01-01 09:00:03    3   Gift // The result
1970-01-01 09:00:04    3   Basket    
```


**Behavior for `backward` and `last_match`**

```SQL
SELECT id, sequenceNextNode('backward', 'last_match')(dt, page = 'Gift', page, page = 'Gift') FROM test_flow GROUP BY id;
                 dt   id   page
1970-01-01 09:00:01    1   Home // The result
1970-01-01 09:00:02    1   Gift // Base point
1970-01-01 09:00:03    1   Exit 
                                     
1970-01-01 09:00:01    2   Home 
1970-01-01 09:00:02    2   Home // The result
1970-01-01 09:00:03    2   Gift // Base point
1970-01-01 09:00:04    2   Basket    
                                     
1970-01-01 09:00:01    3   Gift 
1970-01-01 09:00:02    3   Home // The result
1970-01-01 09:00:03    3   Gift // Base point  
1970-01-01 09:00:04    3   Basket    
```

```SQL
SELECT id, sequenceNextNode('backward', 'last_match')(dt, page = 'Gift', page, page = 'Gift', page = 'Home') FROM test_flow GROUP BY id;
                 dt   id   page
1970-01-01 09:00:01    1   Home // Matched with Home, the result is null
1970-01-01 09:00:02    1   Gift // Base point
1970-01-01 09:00:03    1   Exit 
                                     
1970-01-01 09:00:01    2   Home // The result
1970-01-01 09:00:02    2   Home // Matched with Home
1970-01-01 09:00:03    2   Gift // Base point
1970-01-01 09:00:04    2   Basket    
                                     
1970-01-01 09:00:01    3   Gift // The result
1970-01-01 09:00:02    3   Home // Matched with Home
1970-01-01 09:00:03    3   Gift // Base point  
1970-01-01 09:00:04    3   Basket    
```


**Behavior for `base_condition`**

```SQL
CREATE TABLE test_flow_basecond
(
    `dt` DateTime,
    `id` int,
    `page` String,
    `ref` String
)
ENGINE = MergeTree
PARTITION BY toYYYYMMDD(dt)
ORDER BY id
INSERT INTO test_flow_basecond VALUES (1, 1, 'A', 'ref4') (2, 1, 'A', 'ref3') (3, 1, 'B', 'ref2') (4, 1, 'B', 'ref1');
```

```SQL
SELECT id, sequenceNextNode('forward', 'head')(dt, ref = 'ref1', page, page = 'A') FROM test_flow_basecond GROUP BY id;
                  dt   id   page   ref 
 1970-01-01 09:00:01    1   A      ref4 // The head can't be base point becasue the ref column of the head unmatched with 'ref1'.
 1970-01-01 09:00:02    1   A      ref3 
 1970-01-01 09:00:03    1   B      ref2 
 1970-01-01 09:00:04    1   B      ref1 
 ```
```SQL
SELECT id, sequenceNextNode('backward', 'tail')(dt, ref = 'ref4', page, page = 'B') FROM test_flow_basecond GROUP BY id;
                  dt   id   page   ref 
 1970-01-01 09:00:01    1   A      ref4
 1970-01-01 09:00:02    1   A      ref3 
 1970-01-01 09:00:03    1   B      ref2 
 1970-01-01 09:00:04    1   B      ref1 // The tail can't be base point becasue the ref column of the tail unmatched with 'ref4'.
```

```SQL
SELECT id, sequenceNextNode('forward', 'first_match')(dt, ref = 'ref3', page, page = 'A') FROM test_flow_basecond GROUP BY id;
                  dt   id   page   ref 
 1970-01-01 09:00:01    1   A      ref4 // This row can't be base point becasue the ref column unmatched with 'ref3'.
 1970-01-01 09:00:02    1   A      ref3 // Base point
 1970-01-01 09:00:03    1   B      ref2 // The result
 1970-01-01 09:00:04    1   B      ref1 
```
```SQL
SELECT id, sequenceNextNode('backward', 'last_match')(dt, ref = 'ref2', page, page = 'B') FROM test_flow_basecond GROUP BY id;
                  dt   id   page   ref 
 1970-01-01 09:00:01    1   A      ref4
 1970-01-01 09:00:02    1   A      ref3 // The result
 1970-01-01 09:00:03    1   B      ref2 // Base point
 1970-01-01 09:00:04    1   B      ref1 // This row can't be base point becasue the ref column unmatched with 'ref2'. 
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
```diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+    - backward: Moving backward
+
+-   `base` - Used to set the base point.
+    - head : Set the base point to the first event
```
What actually use-case of `head` and `tail` is? Why we want not to start from first matched event in chain?

achimbab replied:
End-users want to navigate from the beginning of events, but they couldn't know where is the beginning of events. `first_match` and `last_match` has to know where is the beginning. So the `head` and `tail` options may be useful for them.

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
```diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+    
+**Arguments**
+-   `timestamp` — Name of the column containing the timestamp. Data types supported: `Date`, `DateTime` and other unsigned integer types.
+-   `base_condition` — Condition that the base point must fulfill.
```
Why it can be first event from `cond` ?

achimbab replied:
`base_condition` can be moved to the first event-argument with `and combinator`. 

i.e.
```SQL
sequenceNextNode('forward', 'first_match')(date_time, ref_sns = 'facebook', page, page = 'home')
```
The above query can be replaced with the below query.
```SQL
sequenceNextNode('forward', 'first_match')(date_time, page, page = 'home' and ref_sns = 'facebook')
```

I will remove the `base_condition`.

achimbab replied:
`base_condition` can be moved to the first event-argument with `first_match` or `last_match`. 

i.e.
```SQL
sequenceNextNode('forward', 'first_match')(date_time, ref_sns = 'facebook', page, page = 'home')
```
The above query can be replaced with the below query.
```SQL
sequenceNextNode('forward', 'first_match')(date_time, page, page = 'home' and ref_sns = 'facebook')
```

But, `base_condition` is required with `head` and `tail` because there is no way to specify that condition.

vdimir committed:
Minor style fixes in AggregateFunctionSequenceNextNode

vdimir reviewed (commented):
Some questions

achimbab commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
```diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+
+**Parameters**
+-   `direction` - Used to navigate to directions.
+    - forward : Moving forward
```
`forward` with `last_match` and `backward` with `first_match` are useful.
Because `the first matched event` and `the last matched event` will be in the middle of events so that a user want to navigate in both of directions.

`forward` with `tail` and `backward` with `head` is useless.

achimbab commented on src/AggregateFunctions/AggregateFunctionNull.h:
```diff
@@ -276,6 +276,28 @@ class AggregateFunctionNullVariadic final
         this->nested_function->add(this->nestedPlace(place), nested_columns, row_num, arena);
     }
 
+    void insertResultInto(AggregateDataPtr place, IColumn & to, Arena * arena) const override
```
That's right, I will remove the function.

otheng03 committed:
More parameter checking and remove a useless function

achimbab commented:
@vdimir 
I fixed some codes.

- Remove `AggregateFunctionNullVariadic::insertResultInto`
- Prevent `forward` with `tail` and `backward` with `head`

Thank you.

vdimir commented:
@achimbab 

If we consider exception case from  https://github.com/ClickHouse/ClickHouse/pull/19766#issuecomment-795175943


```
CREATE TABLE foo (timestamp DateTime, id Int32, page String) ENGINE=TinyLog;

INSERT INTO foo VALUES (1, 1, 'home'), (2, 1, 'home'), (3, 1, 'giftbox'), (4, 1, 'wish'), (1, 2, 'home'), (2, 2, 'giftbox'), (3, 2, 'wish');
```


```
SELECT
    id,
    sequenceNextNode('forward', 'first_match')(timestamp, page = 'home', page, page = 'home', page = 'giftbox') AS next
FROM foo
GROUP BY id

┌─id─┬─next─┐
│  2 │ wish │
│  1 │ ᴺᵁᴸᴸ │
└────┴──────┘
```

We got with and NULL. But why not wish for id = 1, like

```
1	1	home
2	1	home    // base (first_match)
3	1	giftbox // second event
4	1	wish    // result for id = 1
1	2	home    // base (first_match)
2	2	giftbox // second event
3	2	wish    // result for id = 2

```

achimbab commented:
@vdimir 
| We got with and NULL. But why not wish for id = 1, like

The exceptional case in flow analysis is that:

The number of people decreases from `home` -> `giftbox` but, the number of people increase at `wish`.

```
home (2 people) -> home (1 person)
                -> giftbox (1 person) -> wish (2 people)

// Oops, there are 2 people in the `wish`, but there is only 1 person in the previous flow `giftbox`.
```

So I fixed the algorithm to return NULL for `id = 1` in this case.

```
1	1	home   // base (first_match) because page = 'home'
2	1	home    // not matched, return NULL
3	1	giftbox 
4	1	wish    
```

vdimir commented:
> The number of people decreases from `home` -> `giftbox` but, the number of people increase at `wish`.

I considered semantic of this function close to `windowFunnel` with `strict_order`, but with additional features like `forward`/`backward` and `head`/`tail` and returned next event, but not number of matches.

achimbab commented:
@vdimir 

> Actually, maybe we can consider base as part of chain? Could you provide explicit example why we distinct it?

I will explain why `base_cond` is required for our cases by an example.

**Prepare Data**
```SQL
CREATE TABLE bar (timestamp DateTime, id Int32, page String, ref_url String) ENGINE=TinyLog;
INSERT INTO bar VALUES (1, 1, 'home', ''), (2, 1, 'home', ''), (3, 1, 'giftbox', 'black_friday'), (4, 1, 'wish', ''), (1, 2, 'home', 'black_friday'), (2, 2, 'giftbox', ''), (3, 2, 'wish', ''), (1, 3, 'default', 'black_friday'), (2, 3, 'promotion', '');
```

**Get first pages with ref_url = 'black_friday'**

I want to get the pages where the beginning of rows for each id.

```
SELECT
    id,
    sequenceNextNode('forward', 'head')(timestamp, 1, page)
FROM bar
GROUP BY id
FORMAT PrettySpace

Query id: ebf8eaa9-5b53-4aa0-bb93-c33c6133cfec

 id   sequenceNextNode('forward', 'head')(timestamp, 1, page)

  3   default                                                 
  2   home                                                    
  1   home                                                    
```

It is not achieved with `first_match`, because `first_match` don't return the pages where the beginning of rows for each id.

```SQL
SELECT
    id,
    sequenceNextNode('forward', 'first_match')(timestamp, 1, page, ref_url = 'black_friday')
FROM bar
GROUP BY id
FORMAT PrettySpace

Query id: b787dcc6-c42f-43e1-bb57-7a82cb2a3c4d

 id   sequenceNextNode('forward', 'first_match')(timestamp, 1, page, equals(ref_url, 'black_friday'))

  3   promotion                                                                                       
  2   giftbox          // this value is not the same with the above result.
  1   wish                                                                                            

```

achimbab commented:
@vdimir 

> Actually, maybe we can consider base as part of chain? Could you provide explicit example why we distinct it?

I will explain why `base_cond` is required for our cases by an example.

**Prepare Data**
```SQL
CREATE TABLE bar (timestamp DateTime, id Int32, page String, ref_url String) ENGINE=TinyLog;
INSERT INTO bar VALUES (1, 1, 'home', ''), (2, 1, 'home', ''), (3, 1, 'giftbox', 'black_friday'), (4, 1, 'wish', ''), (1, 2, 'home', 'black_friday'), (2, 2, 'giftbox', ''), (3, 2, 'wish', ''), (1, 3, 'default', 'black_friday'), (2, 3, 'promotion', '');
```

**Get first pages with ref_url = 'black_friday'**

I want to get the pages where the beginning of rows for each id.

```SQL
SELECT
    id,
    sequenceNextNode('forward', 'head')(timestamp, ref_url = 'black_friday', page)
FROM bar
GROUP BY id
FORMAT PrettySpace

Query id: 409f0714-cea5-44bf-a153-8df6008bf9ce

 id   sequenceNextNode('forward', 'head')(timestamp, equals(ref_url, 'black_friday'), page)

  3   default                                                                               
  2   home                                                                                  
  1   ᴺᵁᴸᴸ                                                                                  

```

It is not achieved with `first_match`, because `first_match` don't return the pages where the beginning of rows for each id.

```SQL
SELECT
    id,
    sequenceNextNode('forward', 'first_match')(timestamp, 1, page, ref_url = 'black_friday')
FROM bar
GROUP BY id
FORMAT PrettySpace

Query id: b787dcc6-c42f-43e1-bb57-7a82cb2a3c4d

 id   sequenceNextNode('forward', 'first_match')(timestamp, 1, page, equals(ref_url, 'black_friday'))

  3   promotion                                                                                       
  2   giftbox          // this value is not the same with the above result.
  1   wish             // this value is not the same with the above result.

```

achimbab commented:
If sequenceNextNode('forward', 'first_match')(timestamp, 1, page, ref_url = 'black_friday') returns the current value instead of the next value then sequenceNextNode() requires another argument indicating the position of return value.

vdimir commented:
@achimbab 

Second part of https://github.com/ClickHouse/ClickHouse/pull/19766#issuecomment-814019342

> Is it true that in most cases base_cond and first event will be same? Also order of arguments looks confusing, because we have base_cond between timestamp_column and event_column. Maybe (timestamp, event_column, base_cond, first_cond, second_cond, ...) will be better?

What do you think?


---

Also function still looks a bit special for me, so I'm inclined to merge it available under flag (like `SET allow_experimental_funnel_functions = 1`), so we will able to change signature or remove this function later (maybe once it will be possible to do this with window functions or like that)

otheng03 committed:
Add allow_experimental_funnel_functions and fix argument order of sequenceNextNode

achimbab commented:
@vdimir 

> Also order of arguments looks confusing, because we have base_cond between timestamp_column and event_column. Maybe (timestamp, event_column, base_cond, first_cond, second_cond, ...) will be better?
That's better, I will fix the order of arguments.

> Is it true that in most cases base_cond and first event will be same? 
In most cases end-users will not use `base_cond`, but will use only `first event`. `base_cond` is a filter for `first event`

> Also function still looks a bit special for me, so I'm inclined to merge it available under flag (like SET allow_experimental_funnel_functions = 1), so we will able to change signature or remove this function later (maybe once it will be possible to do this with window functions or like that)
I agree with you.

achimbab commented:
I fixed source codes
- Add allow_experimental_funnel_functions
- Modify the order of arguments

Commit:
Fix typo in Settings.h

vdimir commented on src/AggregateFunctions/AggregateFunctionSequenceNextNode.cpp:
```diff
@@ -0,0 +1,155 @@
+    {
+        if (CurrentThread::isInitialized())
+        {
+            const Context * query_context = CurrentThread::get().getQueryContext();
```
I'll try to merge PR https://github.com/ClickHouse/ClickHouse/pull/22762 first to pass setting explicitly.

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
````diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+**Syntax**
+
+``` sql
+sequenceNextNode(direction, base)(timestamp, base_condition, event_column, event1, event2, event3, ...)
````
```suggestion
sequenceNextNode(direction, base)(timestamp, event_column, base_condition, event1, event2, event3, ...)
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
```diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+    
+**Arguments**
+-   `timestamp` — Name of the column containing the timestamp. Data types supported: `Date`, `DateTime` and other unsigned integer types.
+-   `base_condition` — Condition that the base point must fulfill.
+-   `event_column` — Name of the column containing the value of the next event to be returned. Data types supported: `String` and `Nullable(String)`
```
```suggestion
-   `event_column` — Name of the column containing the value of the next event to be returned. Data types supported: `String` and `Nullable(String)`
-   `base_condition` — Condition that the base point must fulfill.
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
```diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+
+INSERT INTO test_flow VALUES (1, 1, 'A') (2, 1, 'B') (3, 1, 'C') (4, 1, 'E') (5, 1, 'F');
+
+SELECT id, sequenceNextNode('forward', 'head')(dt, page = 'A', page, page = 'A', page = 'B') as next_flow FROM test_flow GROUP BY id;
```
```suggestion
SELECT id, sequenceNextNode('forward', 'head')(dt, page, page = 'A', page = 'A', page = 'B') as next_flow FROM test_flow GROUP BY id;
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
````diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+```
+
+```SQL
+SELECT id, sequenceNextNode('forward', 'head')(dt, page = 'Home', page, page = 'Home', page = 'Gift') FROM test_flow GROUP BY id;
````
```suggestion
SELECT id, sequenceNextNode('forward', 'head')(dt, page, page = 'Home', page = 'Home', page = 'Gift') FROM test_flow GROUP BY id;
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
````diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+**Behavior for `backward` and `tail`**
+
+```SQL
+SELECT id, sequenceNextNode('backward', 'tail')(dt, page = 'Basket', page, page = 'Basket', page = 'Gift') FROM test_flow GROUP BY id;
````
```suggestion
SELECT id, sequenceNextNode('backward', 'tail')(dt, page, page = 'Basket', page = 'Basket', page = 'Gift') FROM test_flow GROUP BY id;
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
````diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+**Behavior for `forward` and `first_match`**
+
+```SQL
+SELECT id, sequenceNextNode('forward', 'first_match')(dt, page = 'Gift', page, page = 'Gift') FROM test_flow GROUP BY id;
````
```suggestion
SELECT id, sequenceNextNode('forward', 'first_match')(dt, page, page = 'Gift', page = 'Gift') FROM test_flow GROUP BY id;
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
````diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+```
+
+```SQL
+SELECT id, sequenceNextNode('forward', 'first_match')(dt, page = 'Gift', page, page = 'Gift', page = 'Home') FROM test_flow GROUP BY id;
````
```suggestion
SELECT id, sequenceNextNode('forward', 'first_match')(dt, page, page = 'Gift', page = 'Gift', page = 'Home') FROM test_flow GROUP BY id;
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
````diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+**Behavior for `backward` and `last_match`**
+
+```SQL
+SELECT id, sequenceNextNode('backward', 'last_match')(dt, page = 'Gift', page, page = 'Gift') FROM test_flow GROUP BY id;
````
```suggestion
SELECT id, sequenceNextNode('backward', 'last_match')(dt, page, page = 'Gift', page = 'Gift') FROM test_flow GROUP BY id;
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
````diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+```
+
+```SQL
+SELECT id, sequenceNextNode('backward', 'last_match')(dt, page = 'Gift', page, page = 'Gift', page = 'Home') FROM test_flow GROUP BY id;
````
```suggestion
SELECT id, sequenceNextNode('backward', 'last_match')(dt, page, page = 'Gift', page = 'Gift', page = 'Home') FROM test_flow GROUP BY id;
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
````diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+```
+
+```SQL
+SELECT id, sequenceNextNode('forward', 'head')(dt, ref = 'ref1', page, page = 'A') FROM test_flow_basecond GROUP BY id;
````
```suggestion
SELECT id, sequenceNextNode('forward', 'head')(dt, page, ref = 'ref1', page = 'A') FROM test_flow_basecond GROUP BY id;
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
````diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+ ```
+
+```SQL
+SELECT id, sequenceNextNode('backward', 'tail')(dt, ref = 'ref4', page, page = 'B') FROM test_flow_basecond GROUP BY id;
````
```suggestion
SELECT id, sequenceNextNode('backward', 'tail')(dt, page, ref = 'ref4', page = 'B') FROM test_flow_basecond GROUP BY id;
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
````diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+```
+
+```SQL
+SELECT id, sequenceNextNode('forward', 'first_match')(dt, ref = 'ref3', page, page = 'A') FROM test_flow_basecond GROUP BY id;
````
```suggestion
SELECT id, sequenceNextNode('forward', 'first_match')(dt, page, ref = 'ref3', page = 'A') FROM test_flow_basecond GROUP BY id;
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
````diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+```
+
+```SQL
+SELECT id, sequenceNextNode('backward', 'last_match')(dt, ref = 'ref2', page, page = 'B') FROM test_flow_basecond GROUP BY id;
````
```suggestion
SELECT id, sequenceNextNode('backward', 'last_match')(dt, page, ref = 'ref2', page = 'B') FROM test_flow_basecond GROUP BY id;
```

vdimir commented on docs/en/sql-reference/aggregate-functions/parametric-functions.md:
```diff
@@ -506,3 +506,254 @@ Solution: Write in the GROUP BY query SearchPhrase HAVING uniqUpTo(4)(UserID) >=
+
+## sequenceNextNode {#sequenceNextNode}
+
+Returns a value of next event that matched an event chain.
```
```suggestion
Returns a value of next event that matched an event chain.

_Experimental function, `SET allow_experimental_funnel_functions = 1` to enable it._
```

vdimir reviewed (approved):
I'll merge it after https://github.com/ClickHouse/ClickHouse/pull/22762 to pass settings explicitly into function

vdimir committed:
Apply suggestions from code review, doc for sequenceNextNode

vdimir committed:
Apply suggestions from code review, doc for sequenceNextNode

otheng03 committed:
Merge remote-tracking branch 'origin' into pr_sequenceNextNode

otheng03 committed:
sequenceNextNode() uses settings passed by creator.

achimbab commented:
@vdimir 
I fixed codes to use explicit settings  #22762.

vdimir committed:
Remove example usage of allow_experimental_funnel_functions from windowFunnel

vdimir committed:
Set allow_experimental_funnel_functions = false by default (not used before)

vdimir committed:
fix style AggregateFunctionWindowFunnel.cpp

gyuton commented:
Internal documentation ticket: DOCSUP-10025.
